name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TastytradePDFExtractor.java

```diff
@@ -0,0 +1,529 @@
+package name.abuchen.portfolio.datatransfer.pdf;
+
+import static name.abuchen.portfolio.datatransfer.ExtractorUtils.checkAndSetFee;
+
+import java.math.BigDecimal;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
+import name.abuchen.portfolio.datatransfer.ExtractorUtils;
+import name.abuchen.portfolio.datatransfer.pdf.PDFParser.Block;
+import name.abuchen.portfolio.datatransfer.pdf.PDFParser.DocumentType;
+import name.abuchen.portfolio.datatransfer.pdf.PDFParser.ParsedData;
+import name.abuchen.portfolio.datatransfer.pdf.PDFParser.Transaction;
+import name.abuchen.portfolio.model.AccountTransaction;
+import name.abuchen.portfolio.model.BuySellEntry;
+import name.abuchen.portfolio.model.Client;
+import name.abuchen.portfolio.model.PortfolioTransaction;
+import name.abuchen.portfolio.money.Money;
+import name.abuchen.portfolio.money.Values;
+import name.abuchen.portfolio.util.OccOsiSymbology;
+
+@SuppressWarnings("nls")
+public class TastytradePDFExtractor extends AbstractPDFExtractor
+{
+    private List<BuySellEntry> m_optionTradesInImport = new ArrayList<>();
+
+    public TastytradePDFExtractor(Client client)
+    {
+        super(client);
+        addBankIdentifier("TASTYWORKS");
+        addBankIdentifier("tastytrade");
+        addBankIdentifier("Apex Clearing Corporation");
+
+        addSummaryStatementBuySellTransaction(); // Purchase and sale ( multiple settlements )
+        addDepotStatementTransaction(); // Securities account transactions ( Settlement account )
+        addDividendTransaction(); // Dividends
+    }
+
+    @Override
+    public String getLabel()
+    {
+        return "tastytrade, Inc.";
+    }
+
+    private void addDividendTransaction()
+    {
+        DocumentType type = new DocumentType("A[\\s]+C[\\s]+C[\\s]+O[\\s]+U[\\s]+N[\\s]+T[\\s]+"
+                        + "S[\\s]+T[\\s]+A[\\s]+T[\\s]+E[\\s]+M[\\s]+E[\\s]+N[\\s]+T");
+        this.addDocumentTyp(type);
+
+        Block firstRelevantLine = new Block("^DIVIDEND[\\s]+[\\d]{2}/[\\d]{2}/[\\d]{2}[\\s]+[A-Z][\\s]+.*$");
+        type.addBlock(firstRelevantLine);
+
+        Transaction<AccountTransaction> pdfTransaction = new Transaction<>();
+        pdfTransaction.subject(() -> {
+            AccountTransaction entry = new AccountTransaction();
+            entry.setType(AccountTransaction.Type.DIVIDENDS);
+            entry.setCurrencyCode("USD");
```

Comment:
Maybe it's better they extract the currency globally from the document instead of setting it manually over and over again.
```
        final DocumentType type = new DocumentType("XXXXXXX", (context, lines) -> {
            Pattern pCurrency = Pattern.compile("^.* XXXXXX, (?<currency>[\\w]{3})$");

            for (String line : lines)
            {
                Matcher mCurrency = pCurrency .matcher(line);
                if (mCurrency.matches())
                    context.put("currency", mCurrency.group("currency"));
                    break;
            }
        });
        this.addDocumentTyp(type);
```
After that you can pick them up in the individual sections
```
        .assign((t, v) -> {
               Map<String, String> context = type.getCurrentContext();
               v.put("currency", context.get("currency"));
               ...
        })
```

Reply:
The currency is always USD. IMHO no change is required.
Is there a way to place it by default in ParsedData?
Is it necessary to always set the currency in the transaction?